### PR TITLE
Suggestion for #118

### PR DESCRIPTION
**Pull request by fmrico**
Hi @jjzapf 

I would like to suggest you the changes in this PR, related to https://github.com/IntelligentRoboticsLabs/ros2_planning_system/pull/118

I hope that you find them interesting!!

Thanks!!

Signed-off-by: Francisco Martín Rico <fmrico@gmail.com>

### CODE REVIEW

File: plansys2_problem_expert/include/plansys2_problem_expert/ProblemExpertClient.hpp [hunk 1]
```diff
@@ -59,36 +59,19 @@ class ProblemExpertClient : public ProblemExpertInterface
   bool removeInstance(const plansys2::Instance & instance);
   std::optional<plansys2::Instance> getInstance(const std::string & name);
 
-  std::vector<plansys2_msgs::msg::Node> getPredicateNodes();
```
Comment (xydesa) [inline plansys2_problem_expert/include/plansys2_problem_expert/ProblemExpertClient.hpp:62]: I think you also want to remove the functions declared on line 52-55 of this file.

File: plansys2_core/include/plansys2_core/Types.hpp [hunk 2]
```diff
@@ -39,39 +39,58 @@ typedef std::vector<PlanItem> Plan;
 class Instance : public plansys2_msgs::msg::Param
 {
 public:
+  Instance()
+  : plansys2_msgs::msg::Param() {}
   explicit Instance(const std::string & name, const std::string & type = {})
   : plansys2_msgs::msg::Param(parser::pddl::fromStringParam(name, type)) {}
-  explicit Instance(const plansys2_msgs::msg::Param & instance)
+  Instance(const plansys2_msgs::msg::Param & instance)
```
Comment (xydesa) [inline plansys2_core/include/plansys2_core/Types.hpp:46]:
```suggestion
  explicit Instance(const plansys2_msgs::msg::Param & instance)
```

File: plansys2_core/include/plansys2_core/Types.hpp [hunk 3]
```diff
@@ -39,39 +39,58 @@ typedef std::vector<PlanItem> Plan;
...
 
 class Predicate : public plansys2_msgs::msg::Node
 {
 public:
+  Predicate()
+  : plansys2_msgs::msg::Node() {}
   explicit Predicate(const std::string & pred)
   : plansys2_msgs::msg::Node(parser::pddl::fromStringPredicate(pred)) {}
-  explicit Predicate(const plansys2_msgs::msg::Node & pred)
+  Predicate(const plansys2_msgs::msg::Node & pred)
```
Comment (xydesa) [inline plansys2_core/include/plansys2_core/Types.hpp:57]:
```suggestion
  explicit Predicate(const plansys2_msgs::msg::Node & pred)
```

File: plansys2_core/include/plansys2_core/Types.hpp [hunk 4]
```diff
@@ -39,39 +39,58 @@ typedef std::vector<PlanItem> Plan;
...
 
 class Function : public plansys2_msgs::msg::Node
 {
 public:
+  Function()
+  : plansys2_msgs::msg::Node() {}
   explicit Function(const std::string & func)
   : plansys2_msgs::msg::Node(parser::pddl::fromStringFunction(func)) {}
-  explicit Function(const plansys2_msgs::msg::Node & func)
+  Function(const plansys2_msgs::msg::Node & func)
```
Comment (xydesa) [inline plansys2_core/include/plansys2_core/Types.hpp:68]:
```suggestion
  explicit Function(const plansys2_msgs::msg::Node & func)
```

File: plansys2_core/include/plansys2_core/Types.hpp [hunk 5]
```diff
@@ -39,39 +39,58 @@ typedef std::vector<PlanItem> Plan;
...
 
 class Goal : public plansys2_msgs::msg::Tree
 {
 public:
+  Goal()
+  : plansys2_msgs::msg::Tree() {}
   explicit Goal(const std::string & goal)
   : plansys2_msgs::msg::Tree(parser::pddl::fromString(goal)) {}
-  explicit Goal(const plansys2_msgs::msg::Tree & goal)
+  Goal(const plansys2_msgs::msg::Tree & goal)
```
Comment (xydesa) [inline plansys2_core/include/plansys2_core/Types.hpp:79]:
```suggestion
  explicit Goal(const plansys2_msgs::msg::Tree & goal)
```
All of the additions of `explicit` to this file in the original PR were inspired by running cpplint on this package via ament_lint_auto:
```
- cpplint
  <<< failure message
    -- run_test.py: invoking following command in '/home/axydes/src/upstream-plansys2-test/src/third_party/plansys2/ros2_planning_system/plansys2_core':
     - /opt/ros/foxy/bin/ament_cpplint --xunit-file /home/axydes/src/upstream-plansys2-test/build/plansys2_core/test_results/plansys2_core/cpplint.xunit.xml
    /home/axydes/src/upstream-plansys2-test/src/third_party/plansys2/ros2_planning_system/plansys2_core/include/plansys2_core/Types.hpp:46:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
    /home/axydes/src/upstream-plansys2-test/src/third_party/plansys2/ros2_planning_system/plansys2_core/include/plansys2_core/Types.hpp:57:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
    /home/axydes/src/upstream-plansys2-test/src/third_party/plansys2/ros2_planning_system/plansys2_core/include/plansys2_core/Types.hpp:68:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
    /home/axydes/src/upstream-plansys2-test/src/third_party/plansys2/ros2_planning_system/plansys2_core/include/plansys2_core/Types.hpp:79:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]
    Category 'runtime/explicit' errors found: 4
    Total errors found: 4
```

**Comment by xydesa** [comment]
In general these changes seem fine to me and work with our use case, once the changes I've suggested have been addressed I think @jjzapf should accept the pull request.

File: plansys2_problem_expert/include/plansys2_problem_expert/ProblemExpertClient.hpp (hunk 1 above)
Comment (jjzapf) [reply plansys2_problem_expert/include/plansys2_problem_expert/ProblemExpertClient.hpp:62]: See commit 25dc11d.

File: plansys2_core/include/plansys2_core/Types.hpp (hunk 2 above)
Comment (jjzapf) [reply plansys2_core/include/plansys2_core/Types.hpp:46]: See commit 25dc11d.

File: plansys2_core/include/plansys2_core/Types.hpp (hunk 3 above)
Comment (jjzapf) [reply plansys2_core/include/plansys2_core/Types.hpp:57]: See commit 25dc11d.

File: plansys2_core/include/plansys2_core/Types.hpp (hunk 4 above)
Comment (jjzapf) [reply plansys2_core/include/plansys2_core/Types.hpp:68]: See commit 25dc11d.

File: plansys2_core/include/plansys2_core/Types.hpp (hunk 5 above)
Comment (jjzapf) [reply plansys2_core/include/plansys2_core/Types.hpp:79]: See commit 25dc11d.

**Comment by jjzapf** [comment]
Thanks @fmrico. The changes you suggested look good and definitely simplify the code. Thank you @xydesa as well for your additional suggestions. I've applied both all suggestions to the pddl-tree-messages branch.

**Pull request merged**